mesonbuild/modules/qt.py

```diff
@@ -0,0 +1,164 @@
+    def preprocess(self, state, args, kwargs):
+        rcc_files, ui_files, moc_headers, moc_sources, sources, include_directories \
+            = extract_as_list(kwargs, 'qresources', 'ui_files', 'moc_headers', 'moc_sources', 'sources', 'include_directories', pop = True)
+        sources += args[1:]
+        method = kwargs.get('method', 'auto')
+        self._detect_tools(state.environment, method)
+        err_msg = "{0} sources specified and couldn't find {1}, " \
+                  "please check your qt{2} installation"
+        if len(moc_headers) + len(moc_sources) > 0 and not self.moc.found():
+            raise MesonException(err_msg.format('MOC', 'moc-qt{}'.format(self.qt_version), self.qt_version))
+        if len(rcc_files) > 0:
+            if not self.rcc.found():
+                raise MesonException(err_msg.format('RCC', 'rcc-qt{}'.format(self.qt_version), self.qt_version))
+            qrc_deps = []
+            for i in rcc_files:
+                qrc_deps += self.parse_qrc(state, i)
+            if len(args) > 0:
+                name = args[0]
+            else:
+                basename = os.path.split(rcc_files[0])[1]
+                name = 'qt' + str(self.qt_version) + '-' + basename.replace('.', '_')
+            rcc_kwargs = {'input': rcc_files,
+                          'output': name + '.cpp',
+                          'command': [self.rcc, '-o', '@OUTPUT@', '@INPUT@'],
+                          'depend_files': qrc_deps}
+            res_target = build.CustomTarget(name, state.subdir, rcc_kwargs)
+            sources.append(res_target)
+        if len(ui_files) > 0:
+            if not self.uic.found():
+                raise MesonException(err_msg.format('UIC', 'uic-qt' + self.qt_version))
+            ui_kwargs = {'output': 'ui_@BASENAME@.h',
+                         'arguments': ['-o', '@OUTPUT@', '@INPUT@']}
+            ui_gen = build.Generator([self.uic], ui_kwargs)
+            ui_output = ui_gen.process_files('Qt{} ui'.format(self.qt_version), ui_files, state)
+            sources.append(ui_output)
+        inc = get_include_args(include_dirs=include_directories)
+        if len(moc_headers) > 0:
+            moc_kwargs = {'output': 'moc_@BASENAME@.cpp',
+                          'arguments': inc + ['@INPUT@', '-o', '@OUTPUT@']}
+            moc_gen = build.Generator([self.moc], moc_kwargs)
+            moc_output = moc_gen.process_files('Qt{} moc header'.format(self.qt_version), moc_headers, state)
+            sources.append(moc_output)
+        if len(moc_sources) > 0:
+            moc_kwargs = {'output': '@BASENAME@.moc',
+                          'arguments': ['@INPUT@', '-o', '@OUTPUT@']}
+            moc_gen = build.Generator([self.moc], moc_kwargs)
+            moc_output = moc_gen.process_files('Qt{} moc source'.format(self.qt_version), moc_sources, state)
+            sources.append(moc_output)
+        return ModuleReturnValue(sources, sources)
+
```

Comment:
__[flake8]__
* [W391] blank line at end of file

<!-- Comment by [SideCI](https://sideci.com) -->